### ISSUE TITLE
Bug in detect return type for get_object_vars.

### ISSUE THREAD

**Issue by mj4444ru** [vimeo/psalm#11215]
```php
$x = new stdClass();
$x->{1} = 1;

var_dump(get_object_vars($x));
```
Return
```
array(1) {
  [1]=>
  int(1)
}
```

https://psalm.dev/r/ae7bc9df71

I'm waiting `array<array-key, mixed>`.

**Comment by mj4444ru** [comment]
> Yeah, this is technically correct but if feels more like an oversight on php implementation than a real feature imho. The ratio between the bugs that would be hidden by loosening this type vs the number of false positive solved feels bad...

PHP has never set restrictions on property names.
You have invented for yourself that a numeric property is an error and you are imposing this on others.

In fact, you need to look at it from the other side. It is not so bad if pslam says that there can be both a string and a number, and the developer can manually correct the type if he is sure that there is always a string. The other side is that you allow the developer to think that there is no error where there may be one. For example, we pass the result of get_object_vars to a function that expects a string as input and we can easily get an elusive error.

**Comment by danog** [comment]
Integer property names are indeed impossible, as php will cast the integer to a string.

The conversion back to an integer happens when the numeric string is used as an array key, due to a nasty behavior of PHP.

I will not support this, because it would mean that array<string, *> is impossible, because theoretically numeric string keys would get casted to integers.

**Comment by mj4444ru** [comment]
> The conversion back to an integer happens when the numeric string is used as an array key, due to a nasty behavior of PHP.

In what form PHP stores the name of the property of the digital format does not matter at all. The fact is that PHP considers such properties normal.


> I will not support this, because it would mean that array<string, *> is impossible, because theoretically numeric string keys would get casted to integers, turning all array<string into array<int|string

I didn't understand what you wanted to say. However, psalm definitely allows you to skip errors now and creates an error where it doesn't actually exist. Everything else is your stereotypes that are harmful to other people.

I have provided exhaustive evidence that your liquidations differ from the standard behavior of PHP and it is very bad that you ignore this.

**Comment by danog** [comment]
Consider the following: https://psalm.dev/r/f05e0218bb

This will throw an exception: https://3v4l.org/s5m4V

In any context, using a numeric string as an array key will automatically cast it to an integer.

I could add an exception specifically for get_object_vars, but that will not cover every other possible case; and since this is a very ugly and specific edge case, I don't see any reason to only add an exception for `get_object_vars`
